index.bs

```diff
@@ -3497,33 +3507,46 @@ IANA "WebAuthn Extension Identifier" registry established by [[!WebAuthn-Registr
+## FIDO AppID Extension (appid) ## {#sctn-appid-extension}
+
+This [=client extension=] allows [=[RPS]=] that have previously registered a
+credential using the legacy FIDO JavaScript APIs to request an assertion. The
+FIDO APIs use an alternative identifier for [=relying parties=] called an |AppID|
+[[FIDO-APPID]], and any credentials created using those APIs will be bound to
+that identifier. Without this extension they would need to be re-registered in
+order to be bound to an [=RP ID=].
 
-This [=authentication extension=] allows [=[RPS]=] that have previously registered a
-credential using the legacy FIDO JavaScript APIs to request an assertion.
-Specifically, this extension allows [=[RPS]=] to specify an |appId| [[FIDO-APPID]]
-to overwrite the otherwise computed |rpId|.  This extension is only valid if
-used during the {{CredentialsContainer/get()}} call; other usage will result in client
-error.
+This extension does not allow FIDO-compatible credentials to be created. Thus
+credentials created with WebAuthn cannot be backwards compatible with the FIDO
+JavaScript APIs.
 
 : Extension identifier
 :: `appid`
 
 : Client extension input
-:: A single JSON string specifying a FIDO |appId|.
+:: A single JSON string specifying a FIDO |AppID|.
 
 : Client extension processing
-:: If {{PublicKeyCredentialRequestOptions/rpId}} is present, return a DOMException
-    whose name is "{{NotAllowedError}}", and terminate this algorithm ([[#discover-from-external-source]]).
-
-    Otherwise, replace the calculation of |rpId| in Step 6 of [[#discover-from-external-source]] with the
-    following procedure:  The client uses the value of |appid| to perform
-    the AppId validation procedure (as defined by [[FIDO-APPID]]).  If valid,
-    the value of |rpId| for all client processing should be replaced by the
-    value of |appid|.
+::  1. If present in a {{CredentialsContainer/create()}} call, return a
+        "{{NotSupportedError}}" {{DOMException}}—this extension is only valid when
+        requesting an assertion.
+    1. Let |facetId| be the result of passing the caller's [=origin=] to the
+        FIDO algorithm for [=determining the FacetID of a calling application=].
+    1. Let |appId| be the extension input.
+    1. Pass |facetId| and |appId| to the FIDO algorithm for [=determining if a
+        caller's FacetID is authorized for an AppID=]. If that algorithm rejects
+        |appId| then return a "{{SecurityError}}" {{DOMException}}.
+    1. When building |allowCredentialDescriptorList|, if a U2F authenticator
+        indicates that a credential is inapplicable (i.e. by returning `SW_WRONG_DATA`)
+        then the client may retry with the U2F application parameter set to the SHA-256
+        hash of |appId|. If this results in an applicable credential, the client may
+        include the credential in |allowCredentialDescriptorList|. The value of |appId|
+        then becomes an additional input parameter for [=authenticatorGetAssertion=] to
```

Comment:
hm. so perhaps we ought to add an optional `appId` param to [authenticatorGetAssertion](https://w3c.github.io/webauthn/#op-get-assertion) ?

Reply:
Based on Emil's comment, I've just specified that `appId` replaces `rpId`. I think this is always going to be a little awkward since we're stitching together two different specs and don't want to burden webauthn with baggage. But implementations have a clear ground-truth here: FIDO credentials should continue to work, so I hope that eliminates any misunderstandings.